File: backend/migrate-remaining-data.py
#!/usr/bin/env python3
"""
Migrate remaining data after partial migration
Skips: categories, products, customers, staff_users
Migrates: orders, order_items, bills, kot_logs
"""

import sqlite3
import psycopg2
import os
import sys

# Database URLs
SQLITE_DB = './data/pos.db'
POSTGRES_URL = os.environ.get('DATABASE_URL')

if not POSTGRES_URL:
    print("❌ ERROR: DATABASE_URL environment variable not set")
    sys.exit(1)

def get_table_columns(cursor, table_name):
    """Get column names from SQLite table"""
    cursor.execute(f"PRAGMA table_info({table_name})")
    return [row[1] for row in cursor.fetchall()]

def migrate_data():
    """Main migration function"""
    print("\n" + "="*80)
    print("MIGRATING REMAINING DATA (Orders, Order Items, Bills, KOT Logs)")
    print("="*80 + "\n")

    try:
        # Connect to both databases
        print("📦 Connecting to databases...")
        sqlite_conn = sqlite3.connect(SQLITE_DB)
        sqlite_conn.row_factory = sqlite3.Row
        sqlite_cursor = sqlite_conn.cursor()

        postgres_conn = psycopg2.connect(POSTGRES_URL, sslmode='require')
        postgres_cursor = postgres_conn.cursor()
        print("✅ Connected to both databases\n")

        # Migration plan
        migration_steps = [
            ("orders", migrate_orders),
            ("order_items", migrate_order_items),
            ("bills", migrate_bills),
            ("kot_logs", migrate_kot_logs),
        ]

        # Execute migrations
        for table_name, migration_func in migration_steps:
            try:
                migration_func(sqlite_cursor, postgres_cursor, postgres_conn)
            except Exception as e:
                print(f"❌ Error migrating {table_name}: {str(e)}")
                postgres_conn.rollback()
                raise

        print("\n" + "="*80)
        print("✅ MIGRATION COMPLETE")
        print("="*80 + "\n")

        # Print summary
        print_summary(sqlite_cursor, postgres_cursor)

    except Exception as e:
        print(f"\n❌ Migration failed: {str(e)}")
        sys.exit(1)
    finally:
        sqlite_conn.close()
        postgres_conn.close()


def migrate_orders(sqlite_cursor, postgres_cursor, postgres_conn):
    """Migrate orders table"""
    print("🔄 Migrating orders...")

    # Get available columns
    columns = get_table_columns(sqlite_cursor, 'orders')
    select_columns = ['id', 'customer_id', 'total_amount', 'status', 'payment_method', 'notes', 'created_at', 'updated_at']
    available_columns = [col for col in select_columns if col in columns]

    sqlite_cursor.execute(f"SELECT {', '.join(available_columns)} FROM orders")
    orders = sqlite_cursor.fetchall()

    for order in orders:
        postgres_cursor.execute("""
            INSERT INTO orders (id, customer_id, total_amount, status, payment_method, notes, created_at, updated_at)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
            ON CONFLICT (id) DO NOTHING
        """, (
            order['id'], order['customer_id'], order['total_amount'],
            order['status'], order['payment_method'], order['notes'], order['created_at'], order['updated_at']
        ))

    postgres_conn.commit()
    print(f"✅ Migrated {len(orders)} orders")


def migrate_order_items(sqlite_cursor, postgres_cursor, postgres_conn):
    """Migrate order_items table"""
    print("🔄 Migrating order items...")

    sqlite_cursor.execute("""
        SELECT id, order_id, product_id, quantity, unit_price, gst, total_price, created_at
        FROM order_items
    """)
    items = sqlite_cursor.fetchall()

    for item in items:
        postgres_cursor.execute("""
            INSERT INTO order_items (id, order_id, product_id, quantity, unit_price, gst, total_price, created_at)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
            ON CONFLICT (id) DO NOTHING
        """, (
            item['id'], item['order_id'], item['product_id'], item['quantity'],
            item['unit_price'], item['gst'], item['total_price'], item['created_at']
        ))

    postgres_conn.commit()
    print(f"✅ Migrated {len(items)} order items")


def migrate_bills(sqlite_cursor, postgres_cursor, postgres_conn):
    """Migrate bills table"""
    print("🔄 Migrating bills...")

    sqlite_cursor.execute("""
        SELECT id, order_id, bill_number, customer_id, total_amount, payment_method, status, notes, created_at, updated_at
        FROM bills
    """)
    bills = sqlite_cursor.fetchall()

    for bill in bills:
        postgres_cursor.execute("""
            INSERT INTO bills (id, order_id, bill_number, customer_id, total_amount, payment_method, status, notes, created_at, updated_at)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            ON CONFLICT (id) DO NOTHING
        """, (
            bill['id'], bill['order_id'], bill['bill_number'], bill['customer_id'],
            bill['total_amount'], bill['payment_method'], bill['status'],
            bill['notes'], bill['created_at'], bill['updated_at']
        ))

    postgres_conn.commit()
    print(f"✅ Migrated {len(bills)} bills")


def migrate_kot_logs(sqlite_cursor, postgres_cursor, postgres_conn):
    """Migrate KOT logs table"""
    print("🔄 Migrating KOT logs...")

    sqlite_cursor.execute("""
        SELECT id, order_id, status, printed_at, completed_at, created_at
        FROM kot_logs
    """)
    kot_logs = sqlite_cursor.fetchall()

    for log in kot_logs:
        postgres_cursor.execute("""
            INSERT INTO kot_logs (id, order_id, status, printed_at, completed_at, created_at)
            VALUES (%s, %s, %s, %s, %s, %s)
            ON CONFLICT (id) DO NOTHING
        """, (
            log['id'], log['order_id'], log['status'], log['printed_at'],
            log['completed_at'], log['created_at']
        ))

    postgres_conn.commit()
    print(f"✅ Migrated {len(kot_logs)} KOT logs")


def print_summary(sqlite_cursor, postgres_cursor):
    """Print migration summary"""
    print("📊 Data Summary:\n")

    tables = ['orders', 'order_items', 'bills', 'kot_logs']

    for table in tables:
        sqlite_cursor.execute(f"SELECT COUNT(*) as count FROM {table}")
        sqlite_count = sqlite_cursor.fetchone()['count']

        postgres_cursor.execute(f"SELECT COUNT(*) as count FROM {table}")
        postgres_count = postgres_cursor.fetchone()[0]

        status = "✅" if sqlite_count == postgres_count else "⚠️"
        print(f"  {status} {table:15} | SQLite: {sqlite_count:5} | PostgreSQL: {postgres_count:5}")


if __name__ == '__main__':
    migrate_data()

File: backend/migrate-remaining-data.js
#!/usr/bin/env node
/**
 * Migrate remaining data after partial migration
 * Skips: categories, products, customers, staff_users
 * Migrates: orders, order_items, bills, kot_logs
 */

import sqlite3 from 'sqlite3';
import pg from 'pg';
import dotenv from 'dotenv';

dotenv.config({ path: '.env' });

const { Pool } = pg;

const SQLITE_DB = './data/pos.db';
const DATABASE_URL = process.env.DATABASE_URL;

if (!DATABASE_URL) {
  console.error('❌ ERROR: DATABASE_URL environment variable not set');
  process.exit(1);
}

class DataMigrator {
  constructor() {
    this.sqliteDb = new sqlite3.Database(SQLITE_DB);
    this.pgPool = new Pool({
      connectionString: DATABASE_URL,
      ssl: { rejectUnauthorized: false }
    });
    this.stats = {
      orders: 0,
      order_items: 0,
      bills: 0,
      kot_logs: 0
    };
  }

  async migrate() {
    console.log('\n' + '='.repeat(80));
    console.log('MIGRATING REMAINING DATA (Orders, Order Items, Bills, KOT Logs)');
    console.log('='.repeat(80) + '\n');

    try {
      console.log('📦 Connecting to databases...');

      // Test connections
      await this.testSQLiteConnection();
      await this.testPostgresConnection();
      console.log('✅ Connected to both databases\n');

      // Migration steps
      await this.migrateOrders();
      await this.migrateOrderItems();
      await this.migrateBills();
      await this.migrateKotLogs();

      console.log('\n' + '='.repeat(80));
      console.log('✅ MIGRATION COMPLETE');
      console.log('='.repeat(80) + '\n');

      this.printSummary();
    } catch (error) {
      console.error('\n❌ Migration failed:', error.message);
      process.exit(1);
    } finally {
      this.close();
    }
  }

  testSQLiteConnection() {
    return new Promise((resolve, reject) => {
      this.sqliteDb.get('SELECT 1', (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  async testPostgresConnection() {
    const client = await this.pgPool.connect();
    await client.query('SELECT 1');
    client.release();
  }

  async migrateOrders() {
    console.log('🔄 Migrating orders...');

    // Get column names from SQLite to handle variations
    const columns = await this.getTableColumns('orders');

    // Build SELECT query with available columns
    const selectColumns = ['id', 'customer_id', 'total_amount', 'status', 'payment_method', 'notes', 'created_at', 'updated_at'];
    const availableColumns = selectColumns.filter(col => columns.includes(col));

    const rows = await this.sqliteQuery(
      `SELECT ${availableColumns.join(', ')} FROM orders`
    );

    for (const row of rows) {
      await this.pgPool.query(
        `INSERT INTO orders (id, customer_id, total_amount, status, payment_method, notes, created_at, updated_at)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
         ON CONFLICT (id) DO NOTHING`,
        [row.id, row.customer_id, row.total_amount, row.status,
         row.payment_method, row.notes, row.created_at, row.updated_at]
      );
    }

    this.stats.orders = rows.length;
    console.log(`✅ Migrated ${rows.length} orders`);
  }

  async migrateOrderItems() {
    console.log('🔄 Migrating order items...');

    const rows = await this.sqliteQuery(
      'SELECT id, order_id, product_id, quantity, unit_price, gst, total_price, created_at FROM order_items'
    );

    for (const row of rows) {
      await this.pgPool.query(
        `INSERT INTO order_items (id, order_id, product_id, quantity, unit_price, gst, total_price, created_at)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
         ON CONFLICT (id) DO NOTHING`,
        [row.id, row.order_id, row.product_id, row.quantity, row.unit_price,
         row.gst, row.total_price, row.created_at]
      );
    }

    this.stats.order_items = rows.length;
    console.log(`✅ Migrated ${rows.length} order items`);
  }

  async migrateBills() {
    console.log('🔄 Migrating bills...');

    const rows = await this.sqliteQuery(
      `SELECT id, order_id, bill_number, customer_id, total_amount, payment_method, status, notes, created_at, updated_at
       FROM bills`
    );

    for (const row of rows) {
      await this.pgPool.query(
        `INSERT INTO bills (id, order_id, bill_number, customer_id, total_amount, payment_method, status, notes, created_at, updated_at)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
         ON CONFLICT (id) DO NOTHING`,
        [row.id, row.order_id, row.bill_number, row.customer_id,
         row.total_amount, row.payment_method, row.status, row.notes, row.created_at, row.updated_at]
      );
    }

    this.stats.bills = rows.length;
    console.log(`✅ Migrated ${rows.length} bills`);
  }

  async migrateKotLogs() {
    console.log('🔄 Migrating KOT logs...');

    const rows = await this.sqliteQuery(
      'SELECT id, order_id, status, printed_at, completed_at, created_at FROM kot_logs'
    );

    for (const row of rows) {
      await this.pgPool.query(
        `INSERT INTO kot_logs (id, order_id, status, printed_at, completed_at, created_at)
         VALUES ($1, $2, $3, $4, $5, $6)
         ON CONFLICT (id) DO NOTHING`,
        [row.id, row.order_id, row.status, row.printed_at, row.completed_at, row.created_at]
      );
    }

    this.stats.kot_logs = rows.length;
    console.log(`✅ Migrated ${rows.length} KOT logs`);
  }

  getTableColumns(tableName) {
    return new Promise((resolve, reject) => {
      this.sqliteDb.all(`PRAGMA table_info(${tableName})`, (err, rows) => {
        if (err) reject(err);
        else resolve(rows.map(row => row.name));
      });
    });
  }

  sqliteQuery(query) {
    return new Promise((resolve, reject) => {
      this.sqliteDb.all(query, (err, rows) => {
        if (err) reject(err);
        else resolve(rows || []);
      });
    });
  }

  printSummary() {
    console.log('📊 Data Summary:\n');
    for (const [table, count] of Object.entries(this.stats)) {
      const padding = ' '.repeat(Math.max(0, 15 - table.length));
      console.log(`  ✅ ${table}${padding} → ${count} records migrated`);
    }
    console.log('\n✨ Remaining data successfully migrated to PostgreSQL!');
  }

  close() {
    this.sqliteDb.close();
    this.pgPool.end();
  }
}

// Run migration
const migrator = new DataMigrator();
migrator.migrate();
